pkg/controller/atlasdeployment/atlasdeployment_controller_test.go

```diff
@@ -758,3 +779,195 @@ func TestUniqueKey(t *testing.T) {
+								ReplicationSpecs: []*mongodbatlas.AdvancedReplicationSpec{
+									{
+										ID:       "789123",
+										ZoneName: "Zone 1",
+										RegionConfigs: []*mongodbatlas.AdvancedRegionConfig{
+											{
+												ProviderName: "AWS",
+												RegionName:   "US_EAST_1",
+												ElectableSpecs: &mongodbatlas.Specs{
+													InstanceSize: "M10",
+													NodeCount:    toptr.MakePtr(3),
+												},
+												Priority: toptr.MakePtr(7),
+											},
+										},
+									},
+								},
+								StateName: "IDLE",
+							}, nil, nil
+						},
+						DeleteFunc: func(projectID string, clusterName string) (*mongodbatlas.Response, error) {
+							return nil, nil
+						},
+					},
+					GlobalClusters: &atlas_mock.GlobalClustersClientMock{
+						GetFunc: func(projectID string, clusterName string) (*mongodbatlas.GlobalCluster, *mongodbatlas.Response, error) {
+							return &mongodbatlas.GlobalCluster{}, nil, nil
+						},
+					},
+					CloudProviderSnapshotBackupPolicies: &atlas_mock.CloudProviderSnapshotBackupPoliciesClientMock{
+						GetFunc: func(projectID string, clusterName string) (*mongodbatlas.CloudProviderSnapshotBackupPolicy, *mongodbatlas.Response, error) {
+							return &mongodbatlas.CloudProviderSnapshotBackupPolicy{
+								ClusterID:             "123789",
+								ClusterName:           deployment.GetDeploymentName(),
+								ReferenceHourOfDay:    toptr.MakePtr(int64(20)),
+								ReferenceMinuteOfHour: toptr.MakePtr(int64(30)),
+								RestoreWindowDays:     toptr.MakePtr(int64(7)),
+								Policies: []mongodbatlas.Policy{
+									{
+										ID: "456987",
+										PolicyItems: []mongodbatlas.PolicyItem{
+											{
+												ID:                "987654",
+												FrequencyInterval: 1,
+												FrequencyType:     "days",
+												RetentionUnit:     "weekly",
+												RetentionValue:    1,
+											},
+										},
+									},
+								},
+								AutoExportEnabled:                 toptr.MakePtr(false),
+								UseOrgAndGroupNamesInExportPrefix: toptr.MakePtr(false),
+							}, nil, nil
+						},
+					},
+				}, nil
+			},
+			IsCloudGovFunc: func() bool {
+				return false
+			},
+			IsSupportedFunc: func() bool {
+				return true
+			},
+		}
+
+		reconciler := &AtlasDeploymentReconciler{
+			ResourceWatcher:             watch.NewResourceWatcher(),
+			Client:                      k8sClient,
+			Log:                         logger,
+			AtlasProvider:               atlasProvider,
+			EventRecorder:               record.NewFakeRecorder(10),
+			ObjectDeletionProtection:    false,
+			SubObjectDeletionProtection: false,
+		}
+
+		result, err := reconciler.Reconcile(
+			context.Background(),
+			ctrl.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: deployment.Namespace,
+					Name:      deployment.Name,
+				},
+			},
+		)
+		assert.NoError(t, err)
+		assert.Equal(t, ctrl.Result{Requeue: false, RequeueAfter: 0}, result)
```

Comment:
Example of what I was mentioning before: This test could be just these lines if we replace `atlasProvider` by a function that calls functions to initialise itself as needed. Way easier to read at a glance.

As the test case might vary depending on how such `atlasProvider` is initialised, we can have it get some simple input parameters or have one per test case with a descriptive name : "normalProjectProvider" vs "brokenProjectProvider", etc. Or `atlasProvider(normalProject())`